### PR TITLE
Connections

### PR DESCRIPTION
**Pull request by stephenmdangelo**
I haven't bothered implementing "ignoreUnknownSignals", as it's just related to how much logging the QML Engine does.

### CODE REVIEW

File: src/qtcore/qml/qml.js [hunk 1]
```diff
@@ -321,6 +321,30 @@ function setupGetterSetter(obj, propName, getter, setter) {
...
+        var params = "";
+        for (var j in item[signalName].parameters) {
+            params += j==0 ? "" : ", ";
+            params += item[signalName].parameters[j].name;
+        }
+        value.src = "(function(" + params + ") { _executionContext = __executionContext;" + value.src + "})";
+        value.isFunction = false;
+        value.compile();
+    }
+    var f = value.eval(objectScope, componentScope);
```
Comment (ChALkeR) [inline src/qtcore/qml/qml.js]: Could you give the variablie `f` a bit more semantic name here? It's not clear what it is.

File: src/modules/QtQml/Connections.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/* Connections */
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: Not required.

File: src/modules/QtQml/Connections.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  baseClass: 'QtObject',
+  constructor: function QMLConnections(meta) {
+    callSuper(this, meta);
+
+    createProperty("QtObject", this, "target");
+    createProperty("bool", this, "ignoreUnknownSignals");
+
+    this.target = this.$parent
+
+    var old_target = this.target;
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: perhaps it should be moved closer to `reconnectTarget`? Also, it should be `let`.

File: src/modules/QtQml/Connections.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    var connections = {}
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: Could you port this to the new module definition syntax? Including properties, etc.

`connections` should probably be `this.$connections` to make setCustomSlot a function in the prototype.

File: src/modules/QtQml/Connections.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    this.$setCustomSlot = function(propName, value, objectScope, componentScope) {
+      connections[propName] = {
+        "value" : value,
+        "objectScope" : objectScope,
+        "componentScope" : componentScope
+      }
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: Missing semicolon. Also, `connections[propName] = { value, objectScope, componentScope };`

File: src/modules/QtQml/Connections.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        if (c._currentConnection && old_target && old_target[i] && typeof old_target[i].disconnect == 'function') {
+          old_target[i].disconnect(c._currentConnection);
+        }
+        c._currentConnection = connectSignal(this.target, i, c.value, c.objectScope, c.componentScope);
+      }
+      old_target = this.target;
+    };
+
+    this.targetChanged.connect(this, reconnectTarget);
+    this.Component.completed.connect(this, reconnectTarget);
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: second `this` is not required, should be `.connect(reconnectTarget)`.

File: src/modules/QtQml/Connections.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    this.$setCustomSlot = function(propName, value, objectScope, componentScope) {
+      connections[propName] = {
+        "value" : value,
+        "objectScope" : objectScope,
+        "componentScope" : componentScope
+      }
+    }
+
+    const reconnectTarget = () => {
+      for(var i in connections) {
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: `for (`

File: src/modules/QtQml/Connections.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        var c = connections[i];
+        if (c._currentConnection && old_target && old_target[i] && typeof old_target[i].disconnect == 'function') {
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: `===`

File: src/modules/QtQml/Connections.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+registerQmlType({
+  module: 'QtQml',
+  name: 'Connections',
+  versions: /.*/,
+  baseClass: 'QtObject',
+  constructor: function QMLConnections(meta) {
+    callSuper(this, meta);
+
+    createProperty("QtObject", this, "target");
+    createProperty("bool", this, "ignoreUnknownSignals");
```
Comment (ChALkeR) [inline src/modules/QtQml/Connections.js]: Could you port this to the new module definition syntax? See [modules/QtQuick/Animation.js](https://github.com/qmlweb/qmlweb/blob/master/src/modules/QtQuick/Animation.js).

**Comment by ChALkeR** [comment]
Mostly LGTM, some style nits.

File: src/modules/QtQml/Connections.js (hunk 5 above)
Comment (stephenmdangelo) [inline src/modules/QtQml/Connections.js]: Huh, who knew?! TIL.

**Comment by stephenmdangelo** [comment]
I believe I've fixed all of those issues now.  Let me know if the port to the new module definition syntax is correct.  I'm guessing there's not any way to have `this.$parent` be the `initialValue` of `target`, as there's not really a `this` to reference at that point.

**Comment by ChALkeR** [comment]
@stephenmdangelo Thanks! Yes, there is no way to default to a dynamic value for now. Perhaps we could add functions support for that (e.g. `target: { type: 'QtObject', initialValue: element => element.$parent }`, but that would block functions being the initial values… Perhaps I need some more time to about that. There are other issues with properties that should be fixed before that, perhaps.

LGTM, merging.

**Comment by ChALkeR** [comment]
Landed in 4aa5ea2...0b40e72.

Thanks!

**Pull request closed without merging**